sys/include/net/netbuf.h

```diff
@@ -0,0 +1,66 @@
+/*
+ * Copyright (C) 2019 HAW Hamburg
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @defgroup    net_netbuf Network buffer abstraction
+ * @ingroup     net
+ * @brief       Provides an abstraction for network stack buffers
+ * @{
+ *
+ * @file
+ * @brief       Network buffer abstraction definition
+ *
+ * @author  Jose I. Alamos <jose.alamos@haw-hamburg.de>
+ */
+#ifndef NET_NETBUF_H
+#define NET_NETBUF_H
+
+#include <stdbool.h>
+#include "iolist.h"
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/**
+ * @brief Netbuf descriptor
+ */
+typedef void netbuf_ctx_t;
+
+/**
+ * @brief   Allocate a network buffer with a given size.
+ *
+ * @note    Supposed to be implemented by the networking module.
+ * @note    @p ctx must be NOT NULL if the function returns a valid pointer.
+ *          If this function returns NULL, the value of @p ctx
+ *          is undefined.
+ *
+ * @param[in]       size    size of the buffer to be allocated.
+ * @param[out]      ctx     allocation context.
+ *
+ * @return          pointer to the buffer.
+ * @return          packet context in @p ctx.
+ * @return          NULL if there's not enough memory for allocation.
+ */
+void *netbuf_alloc(size_t size, netbuf_ctx_t **ctx);
```

Comment:
A pointer should be enough... pointer of pointer doesn't make much sense to me... (yes it comes from the fact that `gnrc_pktbuf_t` is a pointer, but that does not need to be the case.

Reply:
I think the confusion comes from the `netbuf_ctx_t` type.
The context IS a pointer, so I need a pointer to the pointer in order to write the ctx there. I think I will remove the `netbuf_ctx_t` type to make it more clear.

Reply:
I removed the `netbuf_ctx_t` struct in https://github.com/RIOT-OS/RIOT/pull/12741/commits/f21f8e32e755613825e33a76f99f0034cbef3d4f. I also fixed the `psdu` check.

Reply:
But what if the context is e.g. just some numerical value like an index in an array or a file descriptor? Then the stack needs to go through a lot of hassle to make that a pointer of a pointer to that numerical value. On the other hand with a pointer you can just to something like the following (based on your GNRC-implementation)

```C
if (pkt) {
    *((void **)ctx) = pkt;
    return pkt->data;
}
```

the only thing you need to adopt in your implementation for GNRC would be that `netbuf_free()` needs to cast the `pktsnip` appropriately:

```C
/* GNRC implementation of the netbuf_free function */
void netbuf_free(void *netbuf)
{
    gnrc_pktbuf_release(*((gnrc_pktsnip_t **)netbuf));
}
```

Reply:
after discussing this offline with @leandrolanzieri, it might make sense to add a `netbuf_ctx_t` typedef that should be defined by the network stack.

The caller of `netbuf_alloc` has no idea about the type of the `ctx`. This snippet:
```c
if (pkt) {
    *((void **)ctx) = pkt;
    return pkt->data;
}
```
will work in this case because the `ctx` variable is a void pointer. However, for `uint16_t` or complex data structures, we would need to map the `netbuf_ctx_t` to the specific type.

Then we can do something like:
```c
netbuf_ctx_t ctx; /* This could be an uint16_t, pointer or struct depending of the network stack */
void *psdu = netbuf_alloc(size, &ctx);

ind.ctx = &ctx;
```

For GNRC the typedef would be:
```c
typedef gnrc_pktsnip_t* netbuf_ctx_t;
```
For an array-based allocator:
```c
typedef uint16_t netbuf_ctx_t;
```
```

Reply:
> The caller of `netbuf_alloc` has no idea about the type of the `ctx`. This snippet:
> 
> ```c
> if (pkt) {
>     *((void **)ctx) = pkt;
>     return pkt->data;
> }
> ```
> 
> will work in this case because the `ctx` variable is a void pointer. However, for `uint16_t` or complex data structures, we would need to map the `netbuf_ctx_t` to the specific type.

Huh, but currently `ctx` _is_ a `void` pointer. The fact that a stack uses `uint16_t` doesn't change that. It than would do something like 

```C
*((uint16_t *)ctx) = foobar;
```

to set the context, so I don't really see the necessity for a stack-specific typedef. The stack is the only one supposed to touch the context, so they can do the casting internally.

Reply:
> Then we can do something like:
> 
> ```c
> netbuf_ctx_t ctx; /* This could be an uint16_t, pointer or struct depending of the network stack */
> void *psdu = netbuf_alloc(size, &ctx);
> 
> ind.ctx = &ctx;
> ```

This won't work, as `&ctx` points to somewhere in the stack, which, if `ind` is used with IPC might not be the thing you want to point to anymore ;-).

Reply:
> Huh, but currently ctx is a void pointer. The fact that a stack uses uint16_t doesn't change that. It than would do something like

The problem here is that the snippet assumes that someone allocated the proper datatype before calling `netbuf_alloc()`.

So:
```c
*((uint16_t *)ctx) = foobar;
```

Assumes that you did:
```
uint16_t ctx;
netbuf_alloc(size, &ctx);
```
But you don't know if `ctx` is an `uint16_t`, `struct`, etc.

How would you do it then without typedefs nor pointers?

Reply:
> This won't work, as &ctx points to somewhere in the stack, which, if ind is used with IPC might not be the thing you want to point to anymore ;-).

I was assuming there was no IPC calls between the PHY indication and the `netif_recv` function. This restriction is raised if the ctx is always a pointer (a.k.a something is allocated in the heap)

Reply:
> So:
> 
> ```c
> *((uint16_t *)ctx) = foobar;
> ```
> 
> Assumes that you did:
> 
> ```
> uint16_t ctx;
> netbuf_alloc(size, &ctx);
> ```
> 
> But you don't know if `ctx` is an `uint16_t`, `struct`, etc.
> 
> How would you do it then without typedefs nor pointers?

I was more thinking of something along the lines of

```C
void *ctx;
netbuf_alloc(size, ctx);
```

but I now see that this would mean that the the callee needs to store the context somewhere not in stack context...

Reply:
So yeah, having a stack-specific `netbuf_ctx_t` would probably be cleaner. I propose to make it like we did it with `sock`, so `netbuf-types.h` can use the types specified in `netbuf.h` if need be:

```C
/* […] */
#ifdef __cplusplus
extern "C" {
#endif

typedef struct netbuf_ctx netbuf_ctx_t;

/* […] (netbuf function definitions) */

#include "netbuf-types.h"

ifdef __cplusplus
}
#endif

#endif /* NET_NETBUF_H */
/** @} */
```

Reply:
Also maybe make it optional via a sub-module or something, in case a stack doesn't need to use `netbuf_ctx_t`.